replicape: adapt servo0 channels for beagleboard lernel 4.14.108-ti

jannau opened the pull request:
More pwmchips are exported in the sysfs and the correct one is now
pwmchip4. The corect device for the servo pwm channels is
/sys/bus/platform/devices/48302200.pwm
The beagleboard kernel 4.14 and 4.19 contain in addition a commit
which chnages the pwm control directories from 'pwmX' to 'pwm-Y:X'.
Since the commit is not in upstream Linux a config option for the
sysfs path is added. Since beaglebone+replicape is the main user
of the linux PWM support the config defaults to the non-standard
path. Fixes #2514.

Signed-off-by: Janne Grunau <janne-gh@jannau.net>

KevinOConnor commented:
Thanks.  Some comments:

- It would be preferable to avoid a new Kconfig option for this.  Can the linux code just look to see if "/sys/class/pwm/pwmchip%u/pwm%u/%s" exists and if not use "/sys/class/pwm/pwm-%u:%u/%s"?

- If we change SERVO_PINS in replicape.py will it break compatibility with existing installations?  Can we use the existence of a file in /sys to enable the alternate pwm name?  See the setup of `sr_spi_bus` in `replicape.py:__init__()` for an example of this.

-Kevin

jannau commented:
Thanks for the hints. I'm less annoyed with this now. Still marked as draft since I haven't tested it with the old setup (kernel 4.4) and only lightly with kernel 4.14.

pusewicz commented on klippy/extras/replicape.py:
```diff
@@ -205,6 +207,15 @@ def __init__(self, config):
         self.sr_spi.spi_send(self.sr_disabled)
+        if not self.host_mcu.is_fileoutput():
+            #pwm = [x for x in os.listdir('/sys/bus/platform/devices/48302200.pwm/pwm/')
+            #        if x.startswith('pwmchip')]
```
I'd drop these commented out lines.

jannau replied:
yes, I forgot to remove them and they broke the CI due to line length

jannau commented:
tested on `4.4.155-ti-r155` and `4.14.108-ti-r131` and works unmodified on both kernels

KevinOConnor commented:
Thanks.  The linux mcu changes look good to me.  If you want to send that as a separate commit, I'll go ahead and commit it.

I'm not sure I understand the replicape.py changes.  I'm not sure what the filter / lambda / next / iter sequence is trying to do.  If you're looking to read a symbol link, I suggest using `os.readlink()` instead (there's an example of this in scripts/flash_usb.py).  In any case, the os calls need to be wrapped in a try/except block to prevent a failure should the file/directory not exist (I'd say just default to pwmchip0 if it fails).  Also, I'd put all the logic in the servo_pwm class - it's not timing sensitive so no harm in scanning twice.

-Kevin

Szeker commented:
I quickly tried out the modifications by @jannau on my printer with a servo - and it seems working. (Haven't printed yet with the mods, only played with the servo.)
However there seems to be a small issue, which is not necessarily connected to these changes:

During startup, the servo makes a short movement, and temporarily moving out of the defined valid range. In my case this means the servo arm moves to a "negative" angle for a short period of time, before go to the configured "initial_angle".

The same happens, when the servo is disabled -it goes crazy.

My assumption is, that the servo signal buffer enable pin (SERVO_0_EN, enabling the output buffer on the PWM signal going to the servo) is enabled before the PWM pin of the BBB is configured as an output (or before configured with all correct timing values).
That could lead to output a "buffered noise" to the servo, which goes crazy for that random signal.
If the output buffer of the servo signal is disabled, the pull-up resistors on the connector side ensures a stable signal, so the servo is staying stable in it's last position.

I assume, something similar happens, when the servo is disabled by the SET_SERVO... ENABLE=0 command: the output buffer remains enabled while the PWM pin is disabled, which leads to sending out "buffered noise" on the servo control signal.


This is my servo configuration:

pin: replicape:servo0
initial_angle: 20
maximum_servo_angle: 90
minimum_pulse_width: 0.001
maximum_pulse_width: 0.002
enable: True

jannau commented:
@KevinOConnor I splitted the changes in two commits and moved the linux mcu PWM handling to the servo pwm class and added a comment of what the code is trying to do. It is reading the contents of the directory since that directory contains the "real" pwmchipX control directory. The directoryies in `/sys/class/pwm/` which the linux mcu code uses are symlinks to this directory. I think it's useful that the linux mcu code is generic and uses the symlinks in /sys/class/pwm.

Tested on TI kernels 4.4, 4.14 and 4.19. The klipper_pru init script doesn't work for me on 4.14 and the PRU remoteproc indices are 0 and 1 in 4.19.

jannau commented:
@Szeker I think you're reight on enable. The PWM pin should be initialized before it's the pin is switched with the shift register.
The pwm shutdown method is not good fit for the linux mcu code. it only sets the duty cycle to the shutdown/default value. I currently don't see what this value is for the replicape/linux hard_pwm pins.

Please test if this [commit](https://github.com/jannau/klipper/commit/2bdcc8b4c3477da8444db11f00af7928b280135e) fixes the issue on enable.

Szeker commented:
@jannau OK, I'll test the enable issue.
Do I need to recompile/install anything or just restart klipper service after updating replicape.py?

KevinOConnor commented on klippy/extras/replicape.py:
```diff
@@ -108,16 +108,31 @@ def set_digital(self, print_time, value):
+                # to this directory.
+                pwmdev = os.listdir(
+                '/sys/devices/platform/ocp/48302000.epwmss/48302200.pwm/pwm/')
+                self.pwmchip = [pwmchip for pwmchip in pwmdev
+                                if pwmchip.startswith('pwmchip')][0]
```
This does not look correct - self.pwmchip isn't used anywhere and pwmchip in the iterator may overwrite the local pwmchip variable depending on python version.

-Kevin

jannau replied:
indeed, I'll push a fix. thanks

jannau commented:
@Szeker just restarting klipper. I would suggest to open a separate issue for this.

KevinOConnor commented on klippy/extras/replicape.py:
```diff
@@ -108,16 +108,31 @@ def set_digital(self, print_time, value):
+                # to this directory.
+                pwmdev = os.listdir(
+                '/sys/devices/platform/ocp/48302000.epwmss/48302200.pwm/pwm/')
+                pwmchip = [pwmchip for pwmchip in pwmdev
+                           if pwmchip.startswith('pwmchip')][0]
```
I highly advice not using the same local variable as iterator variable - it's confusing, in particular because I think different version of python do different things.  So, I suggest something like:
```
pwmchip = [pc for pc in pwmdev if pc.startswith('pwmchip')][0]
```

Thanks,
-Kevin

jannau replied:
done

KevinOConnor commented:
Thanks.

-Kevin
